### PR TITLE
Fix overlapping extraction grids in GridTiler

### LINKED ISSUE #266: GridTiler with multiple tissue regions uses overlapping extraction grids
**Describe the bug**
The `GridTiler` extractor with a slide with multiple tissue regions (arranged in a particular way, see example) uses overlapping extraction grids.

**To Reproduce**
```python
from histolab.masks import TissueMask
from histolab.slide import Slide
from histolab.tiler import GridTiler
from histolab.data import breast_tissue

_, path = breast_tissue()
s = Slide(path, "pp")
mask = TissueMask()

extractor = GridTiler(
    tile_size=(512, 512),
    level=1,
    check_tissue=True,
    tissue_percent=80.0,
    pixel_overlap=0,
)

extractor.locate_tiles(s, mask)
```
Result:
grid_locate

**Expected behavior**
Every tissue region should have _only one_ grid from which to extract from

**Software (please complete the following information):**
 - OS: macOS 10.15.7 
 - Python Version 3.8.6
 - histolab version: 0.2.5

### PR DESCRIPTION
**Pull request by alessiamarcolini**
fixes #266

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -16,18 +16,20 @@ jobs:
       - name: Download WSI artifacts
         id: load-wsis
         run: |
-          mkdir -p tests/fixtures/external_wsi
-          wget http://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-3.svs -P tests/fixtures/external_wsi
+          mkdir -p tests/fixtures/external-svs
+          wget http://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-3.svs -P tests/fixtures/external-svs
+          mv tests/fixtures/external-svs/CMU-3.svs tests/fixtures/external-svs/cmu-3.svs
```
Comment (ernestoarbitrio) [inline .github/workflows/tests.yml:21]: i think this line can be avoided using the ` -O example.html` option for wget. So after the `-O` option you should pass the actual name you want.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -16,18 +16,20 @@ jobs:
...
         with:
           name: wsi-artifact
-          path: tests/fixtures/external_wsi
+          path: tests/fixtures/external-svs
           retention-days: 1
 
   build:
     needs: wsi-artifacts
     runs-on: ${{ matrix.os }}
+    continue-on-error: true
```
Comment (ernestoarbitrio) [inline .github/workflows/tests.yml:32]: remove this line

File: src/histolab/tiler.py [hunk 3]
```diff
@@ -315,28 +322,75 @@ def _grid_coordinates_from_bbox_coordinates(
...
+                    tile_thumb_coords, binary_mask_region
+                ):
+                    tile_wsi_coords = scale_coordinates(
+                        reference_coords=tile_lvl_coords,
+                        reference_size=slide.level_dimensions(level=self.level),
+                        target_size=slide.level_dimensions(level=0),
+                    )
+                    yield tile_wsi_coords
+
+    def _are_coordinates_within_extraction_mask(
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:360]: respect alphabetical order

File: src/histolab/tiler.py [hunk 4]
```diff
@@ -315,28 +322,75 @@ def _grid_coordinates_from_bbox_coordinates(
...
+        tile_thumb_mask = rectangle_to_mask(
+            dims=binary_mask_region.shape, vertices=tile_thumb_coords
+        )
+
+        tile_in_binary_mask = binary_mask_region & tile_thumb_mask
+
+        tile_area = np.count_nonzero(tile_thumb_mask)
+        tile_in_binary_mask_area = np.count_nonzero(tile_in_binary_mask)
+
+        return tile_in_binary_mask_area / tile_area > 0.8
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:393]: can we use a constant at the top of the file with this (0.8) threshold?

File: tests/integration/test_tiler.py [hunk 5]
```diff
@@ -208,6 +217,24 @@ class DescribeGridTiler:
...
                 "tiles-location-images/kidney-48-5-grid-TM-false-20x20",
             ),
+            pytest.param(
+                EXTERNAL_SVS.CMU_3,
+                BiggestTissueBoxMask(),
+                (512, 512),
+                1,
+                False,
+                "tiles-location-images/external-cmu-3-grid-BTB-false-512x512",
+                marks=pytest.mark.skipif(not on_ci(), reason="To run only on CI"),
```
Comment (ernestoarbitrio) [inline tests/integration/test_tiler.py:227]: ❤️

File: tests/integration/test_tiler.py [hunk 6]
```diff
@@ -233,7 +260,7 @@ def it_locates_tiles_on_the_slide(
         )
 
         # --- Expanding test report with actual and expected images ---
-        expand_tests_report(request, expected=expected_img, actual=tiles_location_img)
+        # expand_tests_report(request, expected=expected_img, actual=tiles_location_img)  # noqa: E800, E501
```
Comment (ernestoarbitrio) [inline tests/integration/test_tiler.py:263]: is this intentional commented out?

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
per comments

File: tests/integration/test_tiler.py (hunk 6 above)
Comment (alessiamarcolini) [reply tests/integration/test_tiler.py:263]: I remember the CI breaking with this uncommented, don't know why. I'll retry

File: src/histolab/tiler.py [hunk 7]
```diff
@@ -48,6 +48,8 @@
 
 logger = logging.getLogger("tiler")
 
+COORDS_WITHIN_EXTRACTION_MASK_THRESH = 0.8
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:51]: mmmm i don't like this name honestly ... it seems too long and truncated ... can we find something more clever?

**Review by ernestoarbitrio — APPROVED** [review]
LGTM : look at the stupid comment please

File: src/histolab/tiler.py (hunk 7 above)
Comment (alessiamarcolini) [reply src/histolab/tiler.py:51]: ehm this is exactly what it is 🤔  help

Comment (ernestoarbitrio) [reply src/histolab/tiler.py:51]: Ok at this point then i would prefer `COORDS_WITHIN_EXTRACTION_MASK_THRESHOLD` ... it's just > 3 chars 😄

File: src/histolab/tiler.py [hunk 8]
```diff
@@ -296,17 +300,59 @@ def tile_size(self, tile_size_: Tuple[int, int]):
...
 
+    def _are_coordinates_within_extraction_mask(
+        self,
+        tile_thumb_coords: CoordinatePair,
+        binary_mask_region: np.ndarray,
+    ) -> bool:
+        """Chack whether the ``tile_thumb_coords`` are inside of ``binary_mask_region``.
+
+        Return True if the tile defined by ``tile_thumb_coords`` is inside of the True
+        areas of ``binary_mask_region`` for at least 80% of its area.
```
Comment (nicolebussola) [inline src/histolab/tiler.py:311]: here I would say `Return True if a fixed portion of the tile defined by ``tile_thumb_coords`` is inside the  area of the ``binary_mask_region``

File: src/histolab/tiler.py [hunk 9]
```diff
@@ -296,17 +300,59 @@ def tile_size(self, tile_size_: Tuple[int, int]):
...
+        tile_thumb_coords : CoordinatePair
+            Coordinates of the tile at thumbnail dimension.
+        binary_mask_region : np.ndarray
+            Binary mask with True inside of the tissue region considered.
+
+        Returns
+        -------
+        bool
+            Whether the tile is inside of the True areas of ``binary_mask_region`` for
+            at least 80% of its area.
```
Comment (nicolebussola) [inline src/histolab/tiler.py:324]: same as before

File: src/histolab/tiler.py [hunk 10]
```diff
@@ -296,17 +300,59 @@ def tile_size(self, tile_size_: Tuple[int, int]):
...
+        tile_in_binary_mask_area = np.count_nonzero(tile_in_binary_mask)
+
+        return (
+            tile_in_binary_mask_area / tile_area > COORDS_WITHIN_EXTRACTION_MASK_THRESH
+        )
+
     def _grid_coordinates_from_bbox_coordinates(
-        self, bbox_coordinates: CoordinatePair, slide: Slide
+        self,
+        bbox_coordinates_lvl: CoordinatePair,
```
Comment (nicolebussola) [inline src/histolab/tiler.py:342]: why ``_lvl``?

File: src/histolab/tiler.py [hunk 11]
```diff
@@ -296,17 +300,59 @@ def tile_size(self, tile_size_: Tuple[int, int]):
...
+        slide: Slide,
+        binary_mask_region: np.ndarray,
     ) -> CoordinatePair:
         """Generate Coordinates at level 0 of grid tiles within a tissue box.
 
         Parameters
         ----------
-        bbox_coordinates: CoordinatePair
+        bbox_coordinates_lvl : CoordinatePair
             Coordinates of the tissue box from which to calculate the coordinates.
```
Comment (nicolebussola) [inline src/histolab/tiler.py:351]: Coordinates of the tissue box from which to calculate the coordinates of....?

File: src/histolab/tiler.py [hunk 12]
```diff
@@ -296,17 +300,59 @@ def tile_size(self, tile_size_: Tuple[int, int]):
...
         slide : Slide
             Slide from which to calculate the coordinates.
+        binary_mask_region : np.ndarray
+            Binary mask corresponding to the region considered
```
Comment (nicolebussola) [inline src/histolab/tiler.py:355]: "the region considered" is not that clear

File: src/histolab/util.py [hunk 13]
```diff
@@ -154,12 +154,45 @@ def regions_from_binary_mask(binary_mask: np.ndarray) -> List[Region]:
...
+def regions_to_binary_mask(regions: List[Region], dims: Tuple[int, int]) -> np.ndarray:
+    """Create a binary mask given a list of ``regions``.
+
+    For each region ``r``, the areas within ``r.coords`` are filled with True, False
+    outside.
+
+    Parameters
+    ----------
+    regions : List[Region]
+        The regions to consider to create the binary mask.
```
Comment (nicolebussola) [inline src/histolab/util.py:174]: The regions to create the binary mask.

**Review by nicolebussola — CHANGES_REQUESTED** [review]
Minor comments

File: src/histolab/tiler.py (hunk 10 above)
Comment (alessiamarcolini) [reply src/histolab/tiler.py:343]: I just changed the naming, they were the coordinates at the level of extraction before as well

File: src/histolab/tiler.py (hunk 12 above)
Comment (alessiamarcolini) [reply src/histolab/tiler.py:356]: ok changed and added a Notes section for clarity

**Review by nicolebussola — APPROVED** [review]
LGTM

**Pull request merged**